File: src/globals.c
#include "globals.h"

uint8_t RESETLEVEL_FLAG;
bool GAMEOVER_FLAG; //triggers a game over
uint8_t BAR_FLAG; //timer for health bar
bool X_FLAG; //true if left or right key is pressed
bool Y_FLAG; //true if up or down key is pressed
uint8_t CHOC_FLAG; //headache timer
uint8_t action; //player sprite array
uint8_t KICK_FLAG; //hit/burn timer
bool GRANDBRULE_FLAG; //If set, player will be "burned" when hit (fireballs)
bool LADDER_FLAG; //True if in a ladder
bool PRIER_FLAG; //True if player is forced into kneestanding because of low ceiling
uint8_t SAUT_FLAG; //6 if free fall or in the middle of a jump, decremented if on solid surface. Must be 0 to initiate a jump.
uint8_t LAST_ORDER; //Last action (kneestand + jump = silent walk)
uint8_t FURTIF_FLAG; //Silent walk timer
bool DROP_FLAG; //True if an object is throwed forward
bool DROPREADY_FLAG;
bool CARRY_FLAG; //true if carrying something (add 16 to player sprite)
bool POSEREADY_FLAG;
uint8_t ACTION_TIMER; //Frames since last action change
//TITUS_sprite sprite; //Player sprite
//TITUS_sprite sprite2; //Secondary player sprite (throwed objects, "hit" when object hits an enemy, smoke when object hits the floor)
uint8_t INVULNERABLE_FLAG; //When non-zero, boss is invulnerable
uint8_t TAPISFLY_FLAG; //When non-zero, the flying carpet is flying
uint8_t CROSS_FLAG; //When non-zero, fall through certain floors (after key down)
uint8_t GRAVITY_FLAG; //When zero, skip object gravity function
uint8_t FUME_FLAG; //Smoke when object hits the floor
const uint8_t *keystate; //Keyboard state
uint8_t YFALL;
bool POCKET_FLAG;
bool PERMUT_FLAG; //If false, there are no animated tiles on the screen?
uint8_t loop_cycle; //Increased every loop in game loop
uint8_t tile_anim; //Current tile animation (0-1-2), changed every 4th game loop cycle
uint8_t BITMAP_X; //Screen offset (X) in tiles
// uint8_t BITMAP_XM; //Point to the left tile in the tile screen (0 to 19)
uint8_t BITMAP_Y; //Screen offset (Y) in tiles
// uint8_t BITMAP_YM; //Point to the top tile in the tile screen (0 to 11)
bool g_scroll_x; //If true, the screen will scroll in X
int16_t g_scroll_x_target; //If scrolling: scroll until player is in this tile (X)
int16_t g_scroll_px_offset;
int16_t XLIMIT; //The engine will not scroll past this tile before the player have crossed the line (X)
bool g_scroll_y; //If true, the screen will scroll in Y
uint8_t g_scroll_y_target; //If scrolling: scroll until player is in this tile (Y)
uint8_t ALTITUDE_ZERO; //The engine will not scroll below this tile before the player have gone below (Y)
uint16_t IMAGE_COUNTER; //Increased every loop in game loop (0 to 0x0FFF)
int8_t SENSX; //1: walk right, 0: stand still, -1: walk left, triggers the ACTION_TIMER if it changes
uint8_t SAUT_COUNT; //Incremented from 0 to 3 when accelerating while jumping, stop acceleration upwards if >= 3
bool NOSCROLL_FLAG;
bool NEWLEVEL_FLAG; //Finish a level
uint8_t BIGNMI_NBR; //Number of bosses that needs to be killed to finish
uint8_t TAUPE_FLAG; //Used for enemies walking and popping up
uint8_t TAPISWAIT_FLAG; //Flying carpet state
uint8_t SEECHOC_FLAG; //Counter when hit
uint8_t BIGNMI_POWER; //Lives of the boss
bool boss_alive; //True if the boss is alive
uint8_t AUDIOMODE;

bool GODMODE; //If true, the player will not interfere with the enemies
bool NOCLIP; //If true, the player will move noclip
bool DISPLAYLOOPTIME; //If true, display loop time in milliseconds

SPRITE sprites[256];

SPRITEDATA spritedata[256];

uint16_t level_code[16];

File: src/globals.zig
//
// Copyright (C) 2008 - 2024 The OpenTitus team
//
// Authors:
// Eirik Stople
// Petr Mrázek
//
// "Titus the Fox: To Marrakech and Back" (1992) and
// "Lagaf': Les Aventures de Moktar - Vol 1: La Zoubida" (1991)
// was developed by, and is probably copyrighted by Titus Software,
// which, according to Wikipedia, stopped buisness in 2005.
//
// OpenTitus is not affiliated with Titus Software.
//
// OpenTitus is  free software; you can redistribute  it and/or modify
// it under the  terms of the GNU General  Public License as published
// by the Free  Software Foundation; either version 3  of the License,
// or (at your option) any later version.
//
// This program is distributed in the hope that it will be useful, but
// WITHOUT  ANY  WARRANTY;  without   even  the  implied  warranty  of
// MERCHANTABILITY or  FITNESS FOR A PARTICULAR PURPOSE.   See the GNU
// General Public License for more details.
//

const std = @import("std");

pub const screen_width = 20;
pub const screen_height = 12;

pub const TEST_ZONE = 4;
pub const MAX_X = 4;
pub const MAX_Y = 12;
pub const MAP_LIMIT_Y = -1;
pub const FIRST_OBJET = 30;
pub const FIRST_NMI = 101;
pub const MAXIMUM_BONUS = 100;
pub const MAXIMUM_ENERGY = 16;
pub const GESTION_X = 40;
pub const GESTION_Y = 20;
pub const MAX_SPEED_DEAD = 20;

pub const TileCoord = i16;
pub const PixelCoord = i16;

pub var RESETLEVEL_FLAG: u8 = 0;
pub var LOSELIFE_FLAG: bool = false;

// triggers a game over
pub var GAMEOVER_FLAG: bool = false;

// timer for health bar
pub var BAR_FLAG: u8 = 0;

// true if left or right key is pressed
pub var X_FLAG: bool = false;

// true if up or down key is pressed
pub var Y_FLAG: bool = false;

// headache timer
pub var CHOC_FLAG: u8 = 0;

// hit/burn timer
pub var KICK_FLAG: u8 = 0;

// If set, player will be "burned" when hit (fireballs)
pub var GRANDBRULE_FLAG: bool = false;

// True if on a ladder
pub var LADDER_FLAG: bool = false;

//True if player is forced into kneestanding because of low ceiling
pub var PRIER_FLAG: bool = false;

//6 if free fall or in the middle of a jump, decremented if on solid surface. Must be 0 to initiate a jump.
pub var SAUT_FLAG: u8 = 0;

//Last action (kneestand + jump = silent walk)
pub var LAST_ORDER: u8 = 0;

//Silent walk timer
pub var FURTIF_FLAG: u8 = 0;

//True if an object is throwed forward
pub var DROP_FLAG: bool = false;

pub var DROPREADY_FLAG: bool = false;

//true if carrying something (add 16 to player sprite)
pub var CARRY_FLAG: bool = false;

pub var POSEREADY_FLAG: bool = false;

//Frames since last action change
pub var ACTION_TIMER: u8 = 0;

//When non-zero, boss is invulnerable
pub var INVULNERABLE_FLAG: u8 = 0;

//When non-zero, fall through certain floors (after key down)
pub var CROSS_FLAG: u8 = 0;

//When zero, skip object gravity function
pub var GRAVITY_FLAG: u8 = 0;

//Smoke when object hits the floor
pub var FUME_FLAG: u8 = 0;

pub var YFALL: u8 = 0;

pub var POCKET_FLAG: bool = false;

//Increased every loop in game loop
pub var loop_cycle: u8 = 0;

//Current tile animation (0-1-2), changed every 4th game loop cycle
pub var tile_anim: u8 = 0;

//Screen offset (X) in tiles
pub var BITMAP_X: TileCoord = 0;

//Screen offset (Y) in tiles
pub var BITMAP_Y: TileCoord = 0;

//If true, the screen will scroll in X
pub var g_scroll_x: bool = false;

pub var g_scroll_px_offset: PixelCoord = 0;

//The engine will not scroll past this tile before the player have crossed the line (X)
pub var XLIMIT: TileCoord = 0;
pub var XLIMIT_BREACHED: bool = false;

//If true, the screen will scroll in Y
pub var g_scroll_y: bool = false;

//If scrolling: scroll until player is in this tile (Y)
pub var g_scroll_y_target: TileCoord = 0;

//The engine will not scroll below this tile before the player have gone below (Y)
pub var ALTITUDE_ZERO: TileCoord = 0;

//Increased every loop in game loop (0 to 0x0FFF)
pub var IMAGE_COUNTER: u16 = 0;

//1: walk right, 0: stand still, -1: walk left, triggers the ACTION_TIMER if it changes
pub var SENSX: i8 = 0;

//Incremented from 0 to 3 when accelerating while jumping, stop acceleration upwards if >= 3
pub var SAUT_COUNT: u8 = 0;

pub var NOSCROLL_FLAG: bool = false;

//Finish a level
pub var NEWLEVEL_FLAG: bool = false;

//Skip a level without recording completion (cheat)
pub var SKIPLEVEL_FLAG: bool = false;

//Used for enemies walking and popping up
pub var TAUPE_FLAG: u8 = 0;

//When non-zero, the flying carpet is flying
pub var TAPISFLY_FLAG: u8 = 0;

//Flying carpet state
pub var TAPISWAIT_FLAG: u8 = 0;

//Counter when hit
pub var SEECHOC_FLAG: u8 = 0;

//Lives of the boss
pub var boss_lives: u8 = 0;

//True if the boss is alive
pub var boss_alive: bool = false;

//If true, the player will not interfere with the enemies
pub var GODMODE: bool = false;

//If true, the player will move noclip
pub var NOCLIP: bool = false;

//If true, display loop time in milliseconds
pub var DISPLAYLOOPTIME: bool = false;
